File: Daily/APRIL/19_04_2563/2563.py
from bisect import bisect_left, bisect_right


class Solution: 
    def countFairPairs(self, nums: list[int], lower: int, upper:int ) -> int:
        n = len(nums)
        count = 0
        nums.sort()

        for i in range(n):
            lower_bound = lower - nums[i]
            upper_bound = upper - nums[i]

            left_index = bisect_left(nums, lower_bound, i + 1)
            right_index = bisect_right(nums, upper_bound, i + 1)

            count += right_index - left_index

        return count 

File: Daily/APRIL/19_04_2563/2563.js
var countFairPairs = function(nums, lowerm upper) {
  const n = nums.length;
  let count = 0;

  nums.sort((a, b) => a- b);

  function binarySearchLeft( arr, target, start) {
    let left = start;
    let right = arr.lenght;
    while( left < right) {
      const mid = Math.floor((left + right) / 2 );
      if (arr[mid] < target) {
        left = mid + 1;
      } else {
        right = mid;
      }

    }
    return left;
  }

  function binarySearchRight(arr, target, start ) {
    let left = start;
    let right = arr.length;
    while( left < right ) {
      const mid = Math.floor(( left + right) / 2);
      if (arr[mid] <= target) {
        left = mid + 1;
      } else {
        right = mid;
      }
    }
    return left;
  }

  for ( let i = 0; i < n; i++) {
    const lowerBound = lower - nums[i];
    const upperBpund = upper - nums[i];

    const leftIndex = binarySearchLeft(nums, lowerBound, i + 1);
    const rightIndex = binarySearchRight(nums, upperBound, i + 1);

    count+= rightIndex - leftIndex;
    
  }
}
